C++ Chrono support

kjbracey opened the pull request:
### Summary of changes <!-- Required -->

Add Chronos versions of time-related APIs to RTOS classes, Ticker-related classes and EventQueue.

`Kernel::Clock`, `HighResClock`, ` LowPowerClock` and `RealTimeClock` act as the Chrono clocks. (C++ standard clocks `system_clock`, `steady_clock` and `high_resolution_clock` are not supported, due to difficulty in consistent retargetting across toolchains).

Deprecate old non-Chrono APIs.

#### Impact of changes <!-- Optional -->

Existing timing code will generate deprecation warnings, but there should be no functional change.

A follow-up PR will clean-up warnings in rest of code base.

#### Migration actions required <!-- Optional -->

Users should migrate to use Chrono-based APIs as directed by the deprecation messages

### Documentation <!-- Required -->

Design document included.

Basic document updates: https://github.com/ARMmbed/mbed-os-5-docs/pull/1302

Basic example updates: https://github.com/ARMmbed/mbed-os-examples-docs_only/pull/112

Design document needs to be reworked into a docs section.

General Chrono background:

* [N2661 - original C++ proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2661.htm)
* [Video: CppCon 2016: Howard Hinnant “A \<chrono\> Tutorial"](https://www.youtube.com/watch?v=P32hvk8b13M)
* [Introduction to std::chrono](http://rachelnertia.github.io/programming/2018/01/07/intro-to-std-chrono/)

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [X] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [X] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@pan-, @bulislaw 


----------------------------------------------------------------------------------------------------------------

kjbracey committed:
Add Chrono support to Kernel and SysTimer

kjbracey committed:
Add Chrono and predicate support to ConditionVariable

kjbracey committed:
Add Chrono support to ThisThread

kjbracey committed:
Add Chrono support to Semaphore

kjbracey committed:
Add Chrono support to Mutex

kjbracey committed:
Thread: minor space optimisation

kjbracey committed:
Add Chrono support to EventFlags

kjbracey committed:
Add Chrono support to Queue,MemoryPool,Mail

kjbracey committed:
Add Chrono support to Ticker et al

kjbracey committed:
Add mbed_chrono.h for handy extensions

kjbracey committed:
Use Chrono in LowPowerTickerWrapper

kjbracey committed:
Add Chrono support to Event/EventQueue

kjbracey committed:
Fix and update tests to use Chronos APIs

bulislaw commented on drivers/Ticker.h:
```diff
@@ -90,7 +93,8 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
 #endif
     void attach(F &&func, float t)
```
Shoulnd't that be deprecated?

bulislaw commented on drivers/Ticker.h:
```diff
@@ -121,8 +125,21 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
+     *  for threads scheduling.
+     *
+     */
+    void attach(Callback<void()> func, TickerClock::duration t);
```
why are we hiding this behind `TickerClock::duration`, wouldn't it be more readable as `std::chrono::microseconds`?

kjbracey replied:
This is possibly the most important design question. It applies both to `TickerClock::duration` and `Kernel::Clock::duration[_u32]`.  

Writing it like this potentially leaves open the door for changing what the API frequency is. I can envisage doing that. For the ticker, it could be done to matching hardware frequency, which would reduce run-time conversions; for the OS it might be part of changing the tick speed for power reasons, and then minimising conversions at run-time again.

Actually doing such a change would not be smooth though - for example if the tick rate was lowered to 100Hz, then you can't just change the `period`, as 

    ThisThread::sleep_for(100ms);

would fail to compile, as you can't implicitly assign `milliseconds` to centiseconds, even if an exact constant like that.

You would need to probably need to change the APIs to be template form to handle rounding if you did that.

We did previously rule that the RTOS tick rate was fixed at milliseconds, and documented that. This is something of a step back from that.

And it is just to provide scope to minimise run-time conversions. We could alternatively lock these APIs and clocks as Chrono milliseconds and microseconds, with commitment to never change (regardless of actual RTOS tick rate or hardware rate), and accept run-time conversion cost.

Oh, there is also a separate symmetry issue, in that time points must be expressed as `TickerClock::time_point`. So whether or not `TickerClock::duration` is fixed, writing it as `TickerClock::duration` just looks neat next to the `time_point`, and makes explicit that this matches the value you'd get by, say, subtracting two `TickerClock::now` values. No conversion needed.

bulislaw replied:
That makes sense, thanks.

bulislaw commented on drivers/Ticker.h:
```diff
@@ -121,8 +125,21 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
+    MBED_DEPRECATED_SINCE("mbed-os-6.0.0", "Pass a chrono duration, not an integer microsecond count. For example use `10ms` rather than `10000`.")
     void attach_us(Callback<void()> func, us_timestamp_t t);
 
+    /** Attach a function to be called by the Ticker, specifying the interval in microseconds
```
You forgot to update class usage example.

bulislaw commented on drivers/TickerClock.h:
```diff
@@ -0,0 +1,87 @@
+ * operators for `->` and conversion to `ticker_data_t *` are provided allowing
+ * TickerClock to be easily substituted in place of a `ticker_data_t *`.
+ */
+struct TickerClock {
```
I noticed you are using struct instead of class. I know it's ok, but are you following some guidelines/convention or just because?

kjbracey replied:
It has no semantic meaning at all, apart from the minor detail of what the default member/base visibility is. You can use it just for that, just to save typing `public:`.

But in my head, I tend to feel happy calling "non-owning" copyable things "structs". "Class" has a vague suggestion of a deeper structure.

I think I apply a similar thing for `class` versus `typename` in templates. "Class" usually hints at something deep - this is likely a fancy object type, "typename" suggests it's likely a primitive type.

But not sure that it's worth having a hard and fast rule here.

kjbracey replied:
I thought it's worth seeing others' view. The [C++ core guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-class) disagree and say 

> C.8: Use class rather than struct if any member is non-public
> Reason
> Readability. To make it clear that something is being hidden/abstracted. This is a useful convention.

ladislas replied:
Fluent C++ has a good take on this as well:

https://www.fluentcpp.com/2017/06/13/the-real-difference-between-struct-class/

Quoting:

> The C++ Core Guidelines
> The above has been inspired by the C++ Core Guideline (which is a great read by the way), in particular the following:
> 
> C.1: Organize related data into structures (structs or classes)
> C.2: Use class if the class has an invariant; use struct if the data members can vary independently
> C.3: Represent the distinction between an interface and an implementation using a class
> C.8: Use class rather than struct if any member is non-public

I like the idea of C.2 and of struct equivalents to bundle, whereas a class can do things.

Also doing a lot of Swift development, structs are value types whereas classes are reference types.  Which tends to agree with C.2 --> use struct for things that can live independently.

I don't mind having a struct do things as well and not just store data.

kjbracey replied:
It's worth noting that in this particular case, there *is* a private member, but it's actually not really hidden, as there are an implicit conversion and `->` operators that give it straight to you.

This really supposed to just be an enhanced `ticker_data *` that is usable as that, but also adds the `Clock` API.

A more simple clock is actually a bit odd anyway, as they usually have *no* members, and only typedefs and a static `now` function. So they technically don't hit C.8, but then they're not a "bundle" either, they're just an interface (to `now`).

bulislaw commented on drivers/Timer.h:
```diff
@@ -92,21 +97,27 @@ class Timer : private NonCopyable<Timer> {
      */
     us_timestamp_t read_high_resolution_us();
 
+    /** Get in a high resolution type the time passed in microseconds.
```
What's the design choice behind leaving `read_us` and `read_ms`?

kjbracey replied:
It's me being soft. I am inclined to remove them, but that forces people to understand how to do `duration_cast<milliseconds>(t.read_duration()).count()` for what is probably a fairly common operation (at least in the sort of glue code where this is used to provide a "HAL" for something using plain ints).

I only went half-way and removed the `float` form, as I didn't want floats to be used casually.

If you're okay with removing `read_us` and `read_ms`, I am too.

bulislaw replied:
I think it's fair to get rid of it now. I don't really like the `read_duration` I'd really like to have it just `read` especially that we are deprecating the float read(). I know it's against our own rules but for the sake of clarity and cleanliness maybe it's worth just replacing them?

kjbracey replied:
I'm also prefer to have something short like `read`.  `read` is generic, so I'm not a huge fan. Read what? Is there something better and short? It's not `now`, because it's not a `time_point`. 

Maybe `elapsed_time`? <strike>That is what `time_point`s have to return their `duration`</strike>. This isn't a time point, with an elapsed real-time since an epoch, but it *does* measure elapsed time like a stop-watch (with start/stop and reset functions). Still a bit long though.

Correction - `time_point`s call theirs `time_since_epoch`. The equivalent for this would be `elapsed_time`.

bulislaw commented on drivers/Timer.h:
```diff
@@ -92,21 +97,27 @@ class Timer : private NonCopyable<Timer> {
+    MBED_DEPRECATED_SINCE("mbed-os-6.0.0", "Floating point operators should normally be avoided for code size. If really needed, you can use `duration<float>(read_duration()).count()`")
     operator float();
 
     /** Get in a high resolution type the time passed in microseconds.
```
From a design point of view, what's the purpose of this function? It's the same as `read_us` but the return type.

kjbracey replied:
Well, it's the strongly-typed Chrono return. You need it when using Chrono APIs. The question is should we deprecate `read_us`? See other comment.

bulislaw commented on platform/source/SysTimer.cpp:
```diff
@@ -29,6 +29,10 @@ extern "C" {
 
+using namespace std::chrono;
+
+const milliseconds deep_sleep_latency{MBED_CONF_TARGET_DEEP_SLEEP_LATENCY};
```
For my information: Any particular reason that's not constexpr?

kjbracey replied:
No, no particular reason. It should be.

You don't *have* to put `constexpr` instead of `const` on **variables**, as const-qualified variables with constant initialisers are permitted in constant expressions, and hence usable in `constexpr` context.

This retains compatibility with the traditional C++ extension over C that `const int x = 4` can be used in constant expressions.

This is different to **function calls**, which can never be used in constant expressions if they're not marked `constexpr`.

But it's good to stick the `constexpr` on the variable as it forces an immediate check that `deep_sleep_latency` is usable in a constant expression - you'd get a compile error if `milliseconds`' constructor wasn't constexpr.

bulislaw commented on platform/source/mbed_os_timer.h:
```diff
@@ -29,32 +30,65 @@ namespace mbed {
 
 #if MBED_CONF_RTOS_PRESENT
-#define OS_TICK_US (1000000 / OS_TICK_FREQ)
+typedef SysTimer<std::ratio<1, OS_TICK_FREQ>> OsTimer;
```
Are you following some kind of standard for using vs typedef?

kjbracey replied:
No. Original code used `typedef`, and I didn't change it.

I'm quite happy changing to `using` - my only niggling reservation is that Eclipse's outline view doesn't recognise it. (https://bugs.eclipse.org/bugs/show_bug.cgi?id=509120)    Parser and auto-complete recognises it fine, but just is annoyingly absent from the file/class outline, unlike the (T) you see for a typedef.

bulislaw reviewed (commented):
Kevin thanks for this awesome change! I started reviewing it, but I quickly realised that it's missing some context. I think we need to have a design document around all the different tickers and clocks and timers in the system: systimer, ostimer, osclock, clock, tickerclock ...more?. I'm personally lost at this point and to figure this out will take me some time and I'm sure other reviewers will be in the same situation. Not to mention people that will like to understand it in the future. I'll keep going through the review, but it would be great if you could create a document describing the clock/timer situation in Mbed especially in context of this changes.

kjbracey committed:
Add RealTimeClock

kjbracey committed:
Add HighResClock and LowPowerClock

kjbracey committed:
mstd_mutex: Add missing Chrono bits

kjbracey committed:
Add Chrono clock design document

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,381 @@
+
+**Basic use of Chrono time literals**
+
+The chrono namespace must be specified in a using declaration to activate the C++14 literals for durations.
```
we should change using with `using`

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,381 @@
+
+`std::high_resolution_clock` represents the clock with the shortest tick period.
+
+`HighResClock` is the best choice for a high-resolution clock, but it will not run during deep sleep. You can use `HighResClock::lock` to keep it running, or otherwise ensure that deep sleep is locked.
```
I have a question regarding the semantic. And if it's already written in stone, please ignore my comment.

With `DeepSleepLock `, I understand that `lock` will prevent the application from entering deep sleep.

In the case of `HighResClock`, `lock` sounds like you're preventing the high resolution clock from running. 

Would the use of `keep_active` or `prevent_deepsleep` or `lock_deepsleep` be more clear for users?

kjbracey replied:
I get your point, but the motivation is to make it generic - it means that `HighResClock` is a C++ _Lockable_. So it could be used with `scoped_lock`/`lock_guard`/`ScopedLock` or similar.

    my_func()
    {
        HighResClock clock; // dummy object - ScopedLock needs an object to work on, regardless of our methods being static
        ScopedLock<XXXClock> lock(clock);
        do stuff();
        if (error) {
             return;
        }
        do more stuff();
    }

The scoped lock automatically cleans up for you on any return.

A Lockable can be positive (locking in or on) or negative (locking out or off), and we do already have "positive"-named ones like `CriticalSectionLock` and `ScopedRamExecutionLock`. The common thing is that you're locking the temporary state (deep sleep off, IRQs off, mutual exclusion), and then returning to the default state (deep sleep on, IRQs on, no mutual exclusion).

So here you're locking on "clock is running", which is the temporary state, then returning to the default "clock is not running" state.

Arguably it's `DeepSleepLock` which should be named more positively - `ShallowSleepLock` or something, indicating it must maintain that power state or higher, rather than indicating the power state you don't want to go to.

(Oh, another naming analogy - Android's "wake locks". They're not called "suspend locks".)

ladislas replied:
I understand! I did not know about C++ Lockable and you're the expert! :)

For my personal knowledge, is that what we're talking about? http://www.cplusplus.com/reference/concept/Lockable/

kjbracey replied:
Yes, that's what I meant, but the reference corrects me; it's only _BasicLockable_, due to absence of the `try_lock`. Adding that would be trivial, but not sure what the point is.

Only thing I can think of is that it would let you do

    scoped_lock<Mutex, HighResClock> lock(m, clock);

You need to be _Lockable_ rather than just _BasicLockable_ if locking more than 1 thing for the multi-lock anti-deadlock algorithm to work.

ladislas replied:
I might be missing something, but:

I understand you can lock a mutex to get exclusive access to a specific resource or section of code and prevent other threads from running this part until the mutex is unlocked. (example: http://cpp.sh/6wcdi)

But for the XXXClock the `lock` doesn't directly refer the the Clock but instead it refers to the DeepSleep state that you want to avoid. 

But the XXXClock is not really a shared resource, is it?

Maybe this part is what I'm missing:

> You need to be Lockable rather than just BasicLockable if locking more than 1 thing for the multi-lock anti-deadlock algorithm to work.

kjbracey replied:
A "lock" can lock anything.   It doesn't need to be a mutex (mutual exclusion). A mutex is only one of the types of things that can be locked.

You can also lock a door, the running of a clock, preventing thread switches, preventing IRQs, staying awake, or keeping the power state above at least some level.

"Locking" the clock means you're using it. Just like "locking" the mutex means you're using it. But only for the mutex does "using it" mean "mutual exclusion", because that's its specific meaning. For the clock "using it" just means "keep it running".

So that example above `scoped_lock<Mutex, HighResClock> lock(m, clock);` means "lock these two objects for the lifetime of this lock object".  (In C++17, you could just write it more simply as `scoped_lock lock(m, clock)`).

Which in this case would ensure mutual exclusion on the mutex - no-one else locking m can run at the same time - and ensuring running of the clock. Two different things, but both handled by the generic "lock" concept.

And "lock the running of the clock" is an abstraction. For LowPowerClock, it's a no-op. For HighResClock it means "lock deep sleep".

That multi-lock includes an ordering algorithm to prevent deadlock. Conceivably someone taking two mutexes could cause deadlock if they kept holding mutex A while waiting for B, but someone else was holding B while waiting for A. It will internally release the others while waiting for one.

In this clock+mutex case, that wouldn't be a deadlock issue, but it would automatically make sure you didn't keep the high-res clock running while waiting for the mutex, so a slight power saving.

Of course, you can do that manually by just writing

     mutex.lock();
     HighResClock::lock();

in that order because you "know"  `mutex.lock()` is a blocking operation but `HighResClock::lock()` isn't, but that wouldn't necessarily be safe for a pair of mutexes.

This whole thing is a bit contrived - I was just trying to think of something that needed _Lockable_ rather than _BasicLockable_.

[See this article](http://howardhinnant.github.io/dining_philosophers.html) for more info.

kjbracey replied:
> But the XXXClock is not really a shared resource, is it?

Oh, yes, it is, but there's no mutual exclusion requirement.  It can be running, and multiple people can be reading its `time_point`s to monitor the time base.

There is internal critical section locking inside `ticker_read` and other APIs to deal with mutual exclusion for the brief duration of API calls, but that's not visible to users.

In API we only need the higher-level "lock it running" concept, which is thus a reference count about how many people are currently relying on it running.

ladislas replied:
Thanks for the detailed explanation ! :) it makes perfect sense now. Being a non-native English speaker (French), `lock` meant something like "prevent others from using it" or "prevent this from running" instead of "I'm using it".

> And "lock the running of the clock" is an abstraction. For LowPowerClock, it's a no-op. For HighResClock it means "lock deep sleep".

As you said earlier, in this case the `DeepSleepLock` is not well named ;)

kjbracey replied:
I think this generalisation of "lock" in computer science may be a relatively recent thing. I'd agree the meaning is a little forced in some cases. :)

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,381 @@
+
+**Prefer absolute time**
+
+Using absolute time aids precision, guaranteeing that there is no drift due to code execution time. This applies to all clocks.
```
This is extremely important. I don't know if it was present in the older documentation but in this case I'm missed it. I had the drifting issue with `sleep_for` but did not know I could fix it with `sleep_until`.

kjbracey replied:
To some extent the issue can be totally self avoiding, _if you're fast enough_. When you wake up at 500ms, if you complete your work before 501ms, then the granularity of RTOS timer ticks means `sleep_for(500ms)` will wake at 1000ms, then 1500ms, etc. You can't drift by less than a tick.  Probably wouldn't work when using the high-res clock though.

Older documentation didn't cover this - this absolute time thing is a bugbear of mine, and I've been progressively adding more and more absolute APIs since 5.0.

This latest version of the PR adds `attach_absolute` to `Timeout` for the first time - that weren't there despite the underlying `TimerEvent` and `ticker_data_t` being based on absolute time since 5.1 or so.

ladislas replied:
>  When you wake up at 500ms, if you complete your work before 501ms, then the granularity of RTOS timer ticks means sleep_for(500ms) will wake at 1000ms, then 1500ms, etc.

in my test I was printing to the serial port. I'll do more testing when this is merged :)

kjbracey replied:
As it so happens, my updated `Timer` example was measuring a print time. You could use that.

Using buffered serial output would help reduce the speed impact of logging. Turn on ` platform.stdio-buffered-serial` if you haven't already.

ladislas replied:
Will do! Thanks!

ladislas replied:
I've checked my experiment again. Here is the basic setup:

- 3 LEDs
- 3 threads
- each thread blinks a specific LED for every xx ms: 500ms, 1000ms and 1500ms respectively

I think the issue doesn't not come from logging but from the fact that events happen at the same time. After around 30 minutes (maybe less but I did not notice before), the blinking is out of sync.

Is there a way to prevent that?

kjbracey replied:
> Is there a way to prevent that?

Well, yes, use absolute time.  You'll only get away with the relative time being "okay if it's fast enough" if there's nothing else contending for resource in the system.

For threads, you could make those threads higher priority, so they at least get to pre-empt any boring other stuff. Otherwise you're stuck waiting for equal priority things to be pre-empted. (And will always defer to higher-priority things).

If you use `Ticker` to do it from interrupt instead of threads, then that's even higher priority, and it will use absolute time additions internally to keep you going at constant period. (But there would still be scope for start-up stagger, as the first call is on a relative time).

Also `EventQueue:call_in` uses absolute addition, IIRC.

Mind you, in a zero-load system, if the only 3 threads doing anything are blinking, I'd be a bit surprised to see them get out of sync. But any other thread printing could cause disruption.

kjbracey replied:
Note that absolute time is already available without this PR - you can do

    uint64_t t = Kernel::get_ms_count();
    for (;;) {
        t += 500;
        ThisThread::sleep_until(t);
    }

If trying to synchronise 3 threads, you'd probably want to pass the "initial time" as a parameter so all 3 get the same one.

ladislas replied:
Fantastic! I'll try that out :)

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,381 @@
+do_operation();
+microseconds elapsed_time = timer.elapsed_time();
+printf("elapsed time = %d us\n", int(elapsed_time.count()));
+// Deep sleep automatically unlocked when timer is destroyed.
```
when timer is destroyed or stoped? or just destroyed?

kjbracey replied:
Read the next line.

ladislas replied:
Yes I've read it and my feeling is that the "if not destroyed" adds another level of complexity. I'd say:

> Deep sleep is automatically unlocked when the timer is stoped or destroyed
>  ⚠️ Remember to stop or destroy the timer when it's not used

kjbracey replied:
Okay, I was thinking more of that specific example, and imagining it being a block of code where the timer was clearly being destroyed, and commenting on that destruction.

How about I make that explicit and put in a function wrapper and stick a comment on the `}`?

ladislas replied:
that sounds good :)

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,381 @@
+#else
+using Clock = LowPowerClock;
+#endif
+Clock::lock(); // will lock deep sleep if high-res clock, else no-op
```
same comment as above about `lock`

ladislas reviewed (commented):
This is an amazing document that clarifies a lot of concept for me! 

I've made a few comments, hope this helps.

bulislaw commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,418 @@
+do_work();
+t = RealTimeClock::to_time_t(woken + 3h);
+printf("Sleeping until %s", ctime(&t));
+ThisThread::sleep_for(3h);
```
You are missing closing the code section here.

bulislaw commented on drivers/HighResClock.h:
```diff
@@ -0,0 +1,113 @@
+ * methods as dummies to support generic code that works with either clock.
+ *
+ * If code is not in a position to use the lock/unlock methods, use
+ * SimpleHighResClock instead.
```
SteadyHighResClock?

kjbracey replied:
Yep, that was previous naming attempt. `Simple` being meant to imply it had some big drawback! "Simple" things always have a downside...

ladislas replied:
"Limited"?

kjbracey replied:
Well, it's not limited. On the contrary, it's unlimited! Just keeps running...

ladislas replied:
Unsafe? In Swift it means that you have to manage everything yourself and take care of cleaning up when you leave.

kjbracey replied:
It's not really that either, in that it wouldn't be possible to clean up in this form. There's no tracking of how many people are using it, or the option to say you're not any more - that would be the `HighResClock`.

ladislas replied:
Hmm :D then I tend to agree with @bulislaw when he says

> Is that really needed? I'm just afraid it it pushes people towards wrong use patterns.

kjbracey replied:
Yeah, I invented it while writing the design document. It's the answer to the question "what's the direct equivalent of `std::chrono::high_resolution_clock`?"

I originally had it as an example in there, then decided I may as well put it in the code, rather than the Markdown.

I can push it back to just being in the Markdown.

bulislaw commented on drivers/HighResClock.h:
```diff
@@ -0,0 +1,113 @@
+ * Unlike LowPowerClock and HighResClock, no lock/unlock methods are provided. Any generic
+ * code that would make use of them should be using one of those classes instead.
+ */
+class SteadyHighResClock {
```
Is that really needed? I'm just afraid it it pushes people towards wrong use patterns.

kjbracey replied:
I don't know. I recognise the danger, but there is presumably a set of devices which doesn't care about deep sleep.

You could equally achieve the same bad effect by just `start()`ing a `Timer` and leaving it running, which is what some subsystems already do.

Can omit it if you prefer.

kjbracey replied:
I've removed `SteadyHighResClock`

bulislaw commented on drivers/LowPowerClock.h:
```diff
@@ -0,0 +1,66 @@
+ * as UnlockedHighPowerClock. No action is required as the lp_ticker
+ * runs continuously.
+ */
+class LowPowerClock {
```
We don't have a steady version?

kjbracey replied:
Well, it is inherently steady. Could have `using SteadyLowPowerClock = LowPowerClock`?

bulislaw commented on drivers/Ticker.h:
```diff
@@ -17,14 +17,17 @@
 namespace mbed {
+
 /**
  * \defgroup drivers_Ticker Ticker class
```
It may be slightly confusing as the docs are now attached to base class, also i'm not sure how well doxygen will deal with that.

kjbracey replied:
Yes, that will need checking, although it was already an inheritance pile anyway. `Timeout` didn't have any public methods of its own, for example.

bulislaw commented on drivers/Ticker.h:
```diff
@@ -54,26 +58,20 @@ namespace mbed {
  */
-class Ticker : public TimerEvent, private NonCopyable<Ticker> {
-
+class TickerBase : public TimerEvent {
```
Can we copy the TickerBase? You also removed the `NonCopyable` from all child classes.

kjbracey replied:
`TimerEvent` is noncopyable, so dropped the specification from all its derived classes. They inherit the noncopyableness.

I guess I should be retaining it for the classes that need it themselves - ie the ones that have destructors. Make explicit that that class must be noncopyable, because it's managing something, not relying on the base class being noncopyable. So `TickerBase` should have it, `Ticker` needn't.

I guess you could restate the non-copyable for documentation purposes on `Ticker`, but then by the same argument you'd want to restate all the inherited public methods too...

bulislaw commented on drivers/Timeout.h:
```diff
@@ -57,14 +57,72 @@ namespace mbed {
  * @endcode
  */
-class Timeout : public Ticker, private NonCopyable<Timeout> {
+class TimeoutBase : public TickerBase {
```
The above docs section with the example is for the Timeout class, it may be confusing and i'm not sure how decent will be the doxygen render. Is TimeoutBase part of the public API? If not we should probably move it out to one of the private directories.

kjbracey replied:
It's a public inheritance, so all its public methods are public, at least. But I guess the class naming and other details needn't be? But I did mention in the design docs the possibility of inheriting from it to make your own `XXXTimeout`. Tending to thing it should be public.

Previously `Timeout` had the extra "take a `ticker_data_t *` constructor serving that purpose, but that was lacking Doxygen.

bulislaw commented on drivers/Timer.h:
```diff
@@ -51,13 +51,9 @@ namespace mbed {
  * @endcode
  */
-class Timer : private NonCopyable<Timer> {
+class TimerBase : private NonCopyable<TimerBase> {
```
Same comments as for the other classes.

bulislaw commented on events/EventQueue.h:
```diff
@@ -358,28 +362,30 @@ class EventQueue : private mbed::NonCopyable<EventQueue> {
-    template <typename F, typename ...Args>
-    int call_in(int ms, Args ...args);
+    template <typename F, typename ...ArgTs>
+    int call_in(duration ms, F f, ArgTs ...args);
```
How did that work before?

kjbracey replied:
This is just Doxygen documentation in a `DOXYGEN_ONLY` block. Doxygen gets the simplified version.

bulislaw commented on rtos/Kernel.h:
```diff
@@ -42,9 +59,66 @@ namespace Kernel {
+ *       std::chrono::milliseconds as its representation, which makes it signed
+ *       and at least 45 bits (so it will be int64_t or equivalent).
+ */
+struct Clock {
```
I'm lost with all the internal clocks SysTimer->OSClock->Kernel::Clock :( do we really need all of them?

kjbracey replied:
I think so?

Real fully-typed singleton clocks:
* `Kernel::Clock` (ms tick clock maintained by RTOS)
* `HighResClock` == `SteadyHighResClock`
* `LowPowerClock`
* `internal::OsClock` (ms tick clock derived from `LowPowerClock` that interfaces to RTOS)

Then under the hood there are the two non-singleton partially-typed pseudo-clocks:
* `TickerDataClock(ticker_data_t *)`  (microsecond clocks)
* `internal::SysTimer(ticker_data_t *)`  (ms tick clocks)

Top 4 need to exist to get the types strongly correct. Arguably `OsClock` and `Kernel::Clock` should always be freely interconvertible, as they should have the same epoch always, but they're still different clocks ("ask the RTOS" and "ask the internal timing code"). Public code should always be doing the former.

I guess I should add something about the internal RTOS bits to the design doc?

bulislaw replied:
Yeah would be good to have it documented so if someone wants/needs to understand all the dependencies and relations they don't need to trace the code.

bulislaw reviewed (commented):
Thanks Kevin this looks good in general, couple of questions in the code. Also one general design/usability comment. You are pushing the three timing function version pattern across the board eg trylock, trylock_for, trylock_until I understand the reasons behind it, but we probably need to do better job around explaining it to the users. Docs team is working on restructuring our handbook to make it more usable and searchable. We should get a solid section about time APIs, architecture and usage patters. We can use various information from this design document and expand it on hints and best practices. Maybe also a guide which time api is right for me. As with all the tickers, timers, timeouts and clocks it's not very clear or easy to figure out.

One more general thing is about the future of the HW tickers. Ages ago we discussed the need for generic API for platform specific tickers. We sort of have the pattern in place with HAL and Drivers, we will just need to make it more generic. How is the rest of the APIs? Do you think the new Clock abstraction is compatible with the arbitrary platform specific clocks?

bulislaw reviewed (dismissed):
I'm generally ok with this. Most of my comments were addressed one way or another. I'm a bit afraid of amount of changes and new things introduced which unfortunately increases the level of complexity. At the same time the Chrono is step in the right direction. I was thinking if we really need all the Ticker/Timeout/Timer(x2) classes and whether we couldn't combine them with the clocks having just methods to achieve similar level of functionality, but it's probably too late and it could have big side effects.
We need to make sure we have solid design documents and handbook info and maybe a blog post to introduce this change. We should not only focus on APIs but examples, api selection, usage patterns and best practices. Kevin's design document is a great start.

rajkan01 committed:
Fix the CI build and unit test issue

rajkan01 committed:
Fix the CI build issue

kjbracey committed:
Chrono test fixes

* Remove ambiguity in single-parameter Queue::put and get
* Fix type problems in RTC test - add missing include
* Don't attempt to use TimerEvent default constructor
* Remove references to Timer::read_duration

kjbracey committed:
More Chrono test fixes

* Timer test - handle removal of Timer(ticker_data_t *)
* Timer test - use Chrono, don't test deprecated methods
* Kernel tick count test - TEST_ASSERT_WITHIN -> TEST_ASSERT_INT_WITHIN
* Mutex test - fix up Chrono changes
* SysTimer test - adapt to SysTimer Chrono changes
* Thread test - use Chrono
* SysTimer - devirtualize destructor

kjbracey committed:
Out-of-line LowPowerTimeout etc, lock

To match Timer etc, make LowPowerTimer et al out-of-line.

Correct LowPowerTimeout to not lock deep sleep.

kjbracey committed:
Durations don't always zero init

In a couple of places I assumed that durations and time_points always
zero-initialised. This is incorrect - they act as if they were their
representation, so integer durations only zero-init when static.

You can zero init (like integers) by having `{}` initialisation.

0xc0170 commented on rtos/Kernel.h:
```diff
@@ -24,15 +24,32 @@
+/* Internal integer-returning function.
+ *
+ * ARM EABI means that `time_point`s do not get returned in registers, so
+ * it's worth having the actual exteernal definition return an integer, and only
```
`exteernal` -> `external`

0xc0170 reviewed (commented):
One typo only, the rest looks fine to me. I like how we start using units and the simplification in ticker drivers.

I'll reread the design doc today

evedon commented on docs/design-documents/platform/clocks/clocks.md:
````diff
@@ -0,0 +1,418 @@
+printf("Printing that took %d us\n", int(print_time.count()));
+```
+
+This example relies on the destruction of the timer to unlock the clock. If using a timer that doesn't get destroyed, you must remember to stop it manually to allow deep sleep to be entered again.
````
Did you mean "This example relies on the destruction of the timer to unlock **deep sleep**"

kjbracey replied:
Yes and no. The clocks have a locking abstraction, as is shown in the next example. If you're using the clock, then you can lock it to make sure it keeps running. For the `HighResClock`, that then locks deep sleep. So it's a 1 step indirect abstraction, and that's what I was thinking of.

However, `Timer` isn't actually implemented on top of a `Clock` (although it could be) - it is actually locking deep sleep directly. It's got a base doing run-time polymorphism as `TimerBase(ticker_data_t)` - it could have been templates like `TimerBase<Clock>`. The first form was minimal structural change, and would lead to smaller code size if using both `Timer` and `LowPowerTimer. The second would be smaller if using only one.

But it's doing the deep sleep lock for the same logical reason to "lock" the clock represented by the `ticker_data_t` to keep running. And it's a contrast to the manual clock locks in the next example.

ladislas replied:
I know I've already asked a lot of questions about that and I'm sorry if I'm a bit slow to get it. Coming back to the documentation after a few weeks, I'm still a little confused.

Here you say(emphasis mine)

>  If you're using the clock, then **you can lock it to make sure it keeps running**.

Which I think I understand correctly.

And [here](https://github.com/ARMmbed/mbed-os/pull/12425/files#diff-64f10ab2be6a2b6f14ac228129e91028R47) you say:

> [high-resolution clock] It does not run in deep sleep, so its users will usually **lock deep sleep**.

In this case, do you **lock deep sleep**:
- to use it (i.e. go into deep sleep) and *make sure it keeps sleeping* even with the high-res clock (that will stop running)
- or actually to **not go** into deep sleep so that your high-res clock can still run?

My guess is the second, but I'd say something like

> you can lock the high resolution clock to disable deep sleep and make sure the ticker keeps running.

Sorry again for getting back at it again :)

kjbracey replied:
> not go into deep sleep so that your high-res clock can still run?

This.

Our terminology around deep sleep locks is backwards. The "lock deep sleep" calls lock out deep sleep from being entered. The high-res clock will stop if deep sleep is entered.

(Android got this right by calling the lock that stops suspend from being entered a "wake lock", not a "suspend lock".)

In computer science lock something while using it, and unlock it when finished. (I guess the metaphor is a toilet, not a car). You then have to figure out what to call that lock. If we'd called it a "shallow sleep lock", things may have been better.

ladislas replied:
Thanks @kjbracey-arm it makes sense, again ;)

Wouldn't mbed os 6.0 be a good time to also change the name of the deep sleep lock to shallow sleep lock? To make it clearer for new comers? Or is this intended for expert users only?

evedon commented on docs/design-documents/platform/clocks/clocks.md:
````diff
@@ -0,0 +1,418 @@
+**Logging times**
+```C++
+auto woken = RealTimeClock::now();
+time_t t = RealTimeClock::to_time_t(rtc_woken);
````
It should be `woken` instead of `rtc_woken`.

evedon commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,418 @@
+
+The above example could be adapted to use absolute wake times with a helper function, guaranteeing a regular 3-hour period, regardless of how long the work takes:
+
+``C++
```
```suggestion
```C++
```

evedon commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,418 @@
+}
+
+auto woken = RealTimeClock::now();
+time_t t = RealTimeClock::to_time_t(rtc_woken);
```
This should also be `woken`

evedon commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,418 @@
+
+In principle, the periods of these clocks could be changed in future, for example to make `HighResClock` match the native hardware rate. Generic code itself may want to select between RTOS and high-resolution clocks which have different periods. To support both cases you may want to write code that always works and compiles regardless of resolution.
+
+The basic rule is that Chrono conversions are implicit if they don't lose precision. So `milliseconds d = t1 - t2;` will work when `t1` and `t2` have millisecond precision or lower. If they were `RealTimeClock::time_point`s measured in seconds, then there would be an automatic multiplcation by 1000 to convert rates. To permit precision-losing conversions, such as dividing by 1000, you need use `d ...
```
```suggestion
The basic rule is that Chrono conversions are implicit if they don't lose precision. So `milliseconds d = t1 - t2;` will work when `t1` and `t2` have millisecond precision or lower. If they were `RealTimeClock::time_point`s measured in seconds, then there would be an automatic multiplication by 1000 to convert rates. To permit precision-losing conversions, such as dividing by 1000, you need use `duration_cast` or `time_point_cast`.
```

evedon commented on drivers/LowPowerTicker.h:
```diff
@@ -40,16 +40,9 @@ namespace mbed {
  */
-class LowPowerTicker : public Ticker, private NonCopyable<LowPowerTicker> {
-
+class LowPowerTicker : public TickerBase {
```
You could remove `#include "platform/NonCopyable.h"`

evedon commented on drivers/LowPowerTimeout.h:
```diff
@@ -36,14 +35,37 @@ namespace mbed {
+    LowPowerClock::time_point scheduled_time() const
     {
-        _function.call();
+        /* Massage from virtual TickerDataClock::time_point used internally to true HighResClock::time_point */
```
This should be: "to true LowPowerClock::time_point"

evedon commented on drivers/RealTimeClock.h:
```diff
@@ -0,0 +1,144 @@
+    /** Initialize the RTC peripheral
+     *
+     * Power up the RTC in preparation for access. This function must be called
+     * before any other RealTimeClock methods ares called. This does not change the state
```
```suggestion
     * before any other RealTimeClock methods are called. This does not change the state
```

evedon commented on drivers/RealTimeClock.h:
```diff
@@ -0,0 +1,144 @@
+    /** Check if the RTC has the time set and is counting
+     *
+     * @retval false The time reported by the RTC is not valid
+     * @retval true  The time has been set the RTC is counting
```
```suggestion
     * @retval true  The time has been set and the RTC is counting
```

evedon commented on drivers/Ticker.h:
```diff
@@ -82,18 +80,21 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
+     *  @deprecated Pass a chrono duration, not a float second count. For example use `10ms` rather than `0.01f`.
      */
 #if defined(__ICCARM__)
+    MBED_DEPRECATED_SINCE("mbed-os-6.0.0", "Pass a chrono duration, not a float second count. For example use `10ms` rather than `0.01f`.")
```
@bulislaw This means that we will have deprecated APIs in 6.0.0 as we are missing code freeze for beta release.

evedon commented on drivers/source/Timer.cpp:
```diff
@@ -15,107 +15,118 @@
-Timer::~Timer()
+TimerBase::~TimerBase()
 {
-    core_util_critical_section_enter();
```
Did you intend to remove the critical section?

kjbracey replied:
Yes, constructors don't (in general) need protection. There can't be multiple users of the object until construction has finished.

evedon commented on rtos/source/Semaphore.cpp:
```diff
@@ -104,43 +108,53 @@ void Semaphore::acquire()
-    bool success = (status == osErrorResource && millisec == 0) ||
-                   (status == osErrorTimeout && millisec != osWaitForever);
+    bool success = (status == osErrorResource && rel_time == rel_time.zero()) ||
+                   (status == osErrorTimeout);
```
I think you need `status == osErrorTimeout  && rel_time <= Kernel::wait_for_u32_max` here.

kjbracey replied:
I believe I was just trying to simplify the "pickiness". Yes, that is the only time "timeout" should occur, but I felt we don't need to be "asserting" that. (And this is code that goes into a release build, not just a developer assert).

"Resource" errors can occur for two different reasons - one is "not available with zero timeout" and the other is "this semaphore handle is invalid". So it's worth trying to catch and make the second scenario fault, rather than swallowing it.

 Does that make sense?

evedon replied:
Yes, I agree, we can simplify the  "pickiness". 
`Mutex::trylock_for` could have the same change then - my comment here was mostly triggered by the difference between the two files.

evedon reviewed (commented):
Some minor comments/typos.

Excellent and very useful documentation. Thank you!

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
````diff
@@ -0,0 +1,418 @@
+### Using the high-resolution clock
+
+**Measuring elapsed time using Timer**
+```C++
````
some example have a line break before the code block, some don't.

when converting to commonmark with pandoc, a new line is added after `...Timer**`.

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,418 @@
+
+This example relies on the destruction of the timer to unlock the clock. If using a timer that doesn't get destroyed, you must remember to stop it manually to allow deep sleep to be entered again.
+
+**Recording high-resolution time points**
```
some example have a line break before the code block, some don't.

when converting to commonmark with pandoc, a new line is added after `...**`.

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
````diff
@@ -0,0 +1,418 @@
+printf("stage 2 time = %d us\n", int(stage2.count()));
+```
+
+**Blinking an LED at 1kHz**
````
some example have a line break before the code block, some don't.

when converting to commonmark with pandoc, a new line is added after `...**`.

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,418 @@
+
+### Using the real-time clock
+
+**Logging times**
```
some example have a line break before the code block, some don't.

when converting to commonmark with pandoc, a new line is added after `...**`.

ladislas commented on docs/design-documents/platform/clocks/clocks.md:
```diff
@@ -0,0 +1,418 @@
+
+**Allowing choice of clock**
+
+As the high-resolution and low-power clocks provide the same API, code can be written generically to work with either clock. For example:
```
some example have a line break before the code block, some don't.

when converting to commonmark with pandoc, a new line is added after `...example:`.

evedon reviewed (approved):
Looks good.
Let's try to get this PR in this week.
